src/query/functions/temporal/linear_regression.go

```diff
@@ -0,0 +1,159 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package temporal
+
+import (
+	"fmt"
+	"math"
+	"time"
+
+	"github.com/m3db/m3/src/query/executor/transform"
+	"github.com/m3db/m3/src/query/ts"
+)
+
+const (
+	// PredictLinearType predicts the value of time series t seconds from now, based on the input series, using simple linear regression.
+	// PredictLinearType should only be used with gauges.
+	PredictLinearType = "predict_linear"
+
+	// DerivType calculates the per-second derivative of the time series, using simple linear regression.
+	// DerivType should only be used with gauges.
+	DerivType = "deriv"
+)
+
+type linearRegressionProcessor struct {
+	duration float64
+}
+
+func (l linearRegressionProcessor) Init(op baseOp, controller *transform.Controller, opts transform.Options) Processor {
+	return &linearRegressionNode{
+		op:         op,
+		controller: controller,
+		timeSpec:   opts.TimeSpec,
+		duration:   l.duration,
+	}
+}
+
+// NewLinearRegressionOp creates a new base temporal transform for linear regression functions
+func NewLinearRegressionOp(args []interface{}, optype string) (transform.Params, error) {
+	var (
+		duration float64
+		ok       bool
+	)
+
+	switch optype {
+	case PredictLinearType:
+		if len(args) != 2 {
+			return emptyOp, fmt.Errorf("invalid number of args for %s: %d", PredictLinearType, len(args))
+		}
+
+		duration, ok = args[1].(float64)
+		if !ok {
+			return emptyOp, fmt.Errorf("unable to cast to scalar argument: %v for %s", args[1], PredictLinearType)
+		}
+
+	case DerivType:
+		if len(args) != 1 {
```

Comment:
you forgot to assign duration here. Bug ?

Reply:
All good- duration is not needed for `deriv`